// @herb-tools/core/src/visitor.ts
// NOTE: This file is generated by the templates/template.rb script and should not
// be modified manually. See /Users/marcoroth/Development/herb-release-0.6.1/templates/javascript/packages/core/src/visitor.ts.erb

import {
  Node,
  DocumentNode,
  LiteralNode,
  HTMLOpenTagNode,
  HTMLCloseTagNode,
  HTMLElementNode,
  HTMLAttributeValueNode,
  HTMLAttributeNameNode,
  HTMLAttributeNode,
  HTMLTextNode,
  HTMLCommentNode,
  HTMLDoctypeNode,
  XMLDeclarationNode,
  CDATANode,
  WhitespaceNode,
  ERBContentNode,
  ERBEndNode,
  ERBElseNode,
  ERBIfNode,
  ERBBlockNode,
  ERBWhenNode,
  ERBCaseNode,
  ERBCaseMatchNode,
  ERBWhileNode,
  ERBUntilNode,
  ERBForNode,
  ERBRescueNode,
  ERBEnsureNode,
  ERBBeginNode,
  ERBUnlessNode,
  ERBYieldNode,
  ERBInNode,
} from "./nodes.js"

export class Visitor {
  visit(node: Node | null | undefined): void {
    if (!node) return

    node.accept(this)
  }

  visitAll(nodes: (Node | null | undefined)[]): void {
    nodes.forEach(node => node?.accept(this))
  }

  visitChildNodes(node: Node): void {
    node.compactChildNodes().forEach(node => node.accept(this))
  }

  visitDocumentNode(node: DocumentNode): void {
    this.visitChildNodes(node)
  }

  visitLiteralNode(node: LiteralNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLCloseTagNode(node: HTMLCloseTagNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLElementNode(node: HTMLElementNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLAttributeValueNode(node: HTMLAttributeValueNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLAttributeNameNode(node: HTMLAttributeNameNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLAttributeNode(node: HTMLAttributeNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLTextNode(node: HTMLTextNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLCommentNode(node: HTMLCommentNode): void {
    this.visitChildNodes(node)
  }

  visitHTMLDoctypeNode(node: HTMLDoctypeNode): void {
    this.visitChildNodes(node)
  }

  visitXMLDeclarationNode(node: XMLDeclarationNode): void {
    this.visitChildNodes(node)
  }

  visitCDATANode(node: CDATANode): void {
    this.visitChildNodes(node)
  }

  visitWhitespaceNode(node: WhitespaceNode): void {
    this.visitChildNodes(node)
  }

  visitERBContentNode(node: ERBContentNode): void {
    this.visitChildNodes(node)
  }

  visitERBEndNode(node: ERBEndNode): void {
    this.visitChildNodes(node)
  }

  visitERBElseNode(node: ERBElseNode): void {
    this.visitChildNodes(node)
  }

  visitERBIfNode(node: ERBIfNode): void {
    this.visitChildNodes(node)
  }

  visitERBBlockNode(node: ERBBlockNode): void {
    this.visitChildNodes(node)
  }

  visitERBWhenNode(node: ERBWhenNode): void {
    this.visitChildNodes(node)
  }

  visitERBCaseNode(node: ERBCaseNode): void {
    this.visitChildNodes(node)
  }

  visitERBCaseMatchNode(node: ERBCaseMatchNode): void {
    this.visitChildNodes(node)
  }

  visitERBWhileNode(node: ERBWhileNode): void {
    this.visitChildNodes(node)
  }

  visitERBUntilNode(node: ERBUntilNode): void {
    this.visitChildNodes(node)
  }

  visitERBForNode(node: ERBForNode): void {
    this.visitChildNodes(node)
  }

  visitERBRescueNode(node: ERBRescueNode): void {
    this.visitChildNodes(node)
  }

  visitERBEnsureNode(node: ERBEnsureNode): void {
    this.visitChildNodes(node)
  }

  visitERBBeginNode(node: ERBBeginNode): void {
    this.visitChildNodes(node)
  }

  visitERBUnlessNode(node: ERBUnlessNode): void {
    this.visitChildNodes(node)
  }

  visitERBYieldNode(node: ERBYieldNode): void {
    this.visitChildNodes(node)
  }

  visitERBInNode(node: ERBInNode): void {
    this.visitChildNodes(node)
  }

}
